akka-projection-kafka/src/main/scala/akka/projection/kafka/internal/KafkaProjectionImpl.scala

```diff
@@ -0,0 +1,123 @@
+    override val projectionId: ProjectionId,
+    sourceProvider: SourceProvider[Offset, Envelope],
+    offsetStore: OffsetStore[Offset, Future], // Kafka doesn't have its own OffsetStore yet
+    settings: ProducerSettings[Key, Value],
+    strategy: KafkaProjectionImpl.Strategy,
+    handler: Envelope => ProducerRecord[Key, Value])
+    extends Projection[Envelope] {
+  import KafkaProjectionImpl._
+
+  private val killSwitch = KillSwitches.shared(projectionId.id)
+  private val promiseToStop: Promise[Done] = Promise()
+  private val started = new AtomicBoolean(false)
+
+  override def run()(implicit systemProvider: ClassicActorSystemProvider): Unit = {
+    if (started.compareAndSet(false, true)) {
+      implicit val system: ActorSystem = systemProvider.classicSystem
+
+      val done = mappedSource().runWith(Sink.ignore)
+      promiseToStop.completeWith(done)
+    }
+  }
+
+  override def stop()(implicit ec: ExecutionContext): Future[Done] = {
+    if (started.get()) {
+      killSwitch.shutdown()
+      promiseToStop.future
+    } else {
+      Future.failed(new IllegalStateException(s"Projection [$projectionId] not started yet!"))
+    }
+  }
+
+  override private[projection] def mappedSource()(
+      implicit systemProvider: ClassicActorSystemProvider): Source[Done, _] = {
+    implicit val ec = systemProvider.classicSystem.dispatcher
+
+    val lastKnownOffset: Future[Option[Offset]] = offsetStore.readOffset(projectionId)
+
+    val source: Source[(Offset, Envelope), NotUsed] =
+      Source
+        .futureSource(lastKnownOffset.map(sourceProvider.source))
+        .via(killSwitch.flow)
+        .map(envelope => sourceProvider.extractOffset(envelope) -> envelope)
+        .mapMaterializedValue(_ => NotUsed)
+
+    val sendProducer = SendProducer(settings)(systemProvider.classicSystem)
+
+    val handlerFlow: Flow[(Offset, Envelope), Offset, NotUsed] =
+      Flow[(Offset, Envelope)].mapAsync(parallelism = 1) {
+        case (offset, envelope) =>
+          val pr = handler(envelope)
+          val metadata: Future[RecordMetadata] = sendProducer.send(pr)
+          metadata.map(_ => offset)
+      }
+
+    val composedSource: Source[Done, NotUsed] = strategy match {
+      case AtLeastOnce(1, _) =>
+        source
+          .via(handlerFlow)
+          .mapAsync(1) { offset => offsetStore.saveOffsetAsync(projectionId, offset) }
+      case AtLeastOnce(afterEnvelopes, orAfterDuration) =>
+        source
+          .via(handlerFlow)
+          .groupedWithin(afterEnvelopes, orAfterDuration)
+          .collect { case grouped if grouped.nonEmpty => grouped.last }
+          .mapAsync(parallelism = 1) { offset => offsetStore.saveOffsetAsync(projectionId, offset) }
+      case AtMostOnce =>
+        source
+          .mapAsync(parallelism = 1) {
+            case (offset, envelope) => offsetStore.saveOffsetAsync(projectionId, offset).map(_ => offset -> envelope)
+          }
+          .via(handlerFlow)
+          .map(_ => Done)
+    }
```

Comment:
There is something backwards with this compared to what we have done so far. CassandraProjection and SlickProjection are about where the offset is stored. I thought those would be able to store Kafka offsets also (a generalized representation of such offset to avoid the Kafka dependency in those modules).

Then the akka-projection-kafka would provide a `KafkaSourceProvider` with similarities to the `EventSourcedProvider`.

Reply:
I created an example to illustrate what I mean: https://github.com/akka/akka-projection/pull/81/files#r417199141

Reply:
+1

Reply:
I've merged the example into this PR and fleshed it out some more using the Slick impl.